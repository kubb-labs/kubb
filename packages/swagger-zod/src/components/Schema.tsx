import { Oas } from '@kubb/plugin-oas/components'
import { Const, File, useApp, useFile } from '@kubb/react'
import { pluginTsName } from '@kubb/swagger-ts'

import transformers from '@kubb/core/transformers'
import { isKeyword, schemaKeywords } from '@kubb/plugin-oas'
import { useSchema } from '@kubb/plugin-oas/hooks'
import type { ReactNode } from 'react'
import * as parserZod from '../parser/index.ts'
import { pluginZodName } from '../plugin.ts'
import type { PluginZod } from '../types.ts'

type Props = {
  description?: string
  withTypeAnnotation?: boolean
  keysToOmit?: string[]
}

export function Schema(props: Props): ReactNode {
  const { keysToOmit, withTypeAnnotation, description } = props
  const { tree, name } = useSchema()
  const {
    pluginManager,
    plugin: {
      options: { mapper, coercion },
    },
  } = useApp<PluginZod>()

  // all checks are also inside this.schema(React)
  const resolvedName = pluginManager.resolveName({
    name,
    pluginKey: [pluginZodName],
    type: 'function',
  })

  const typeName = pluginManager.resolveName({
    name,
    pluginKey: [pluginTsName],
    type: 'type',
  })

  if (!tree.length) {
    return (
      <Const
        name={resolvedName}
        export
        JSDoc={{
          comments: [description ? `@description ${transformers.jsStringEscape(description)}` : undefined].filter(Boolean),
        }}
      >
        undefined
      </Const>
    )
  }

  const hasTuple = tree.some((item) => isKeyword(item, schemaKeywords.tuple))

  const output = parserZod
    .sort(tree)
    .filter((item) => {
      if (hasTuple && (isKeyword(item, schemaKeywords.min) || isKeyword(item, schemaKeywords.max))) {
        return false
      }

      return true
    })
    .map((item) => parserZod.parse(undefined, item, { name, typeName, description, mapper, coercion, keysToOmit }))
    .filter(Boolean)
    .join('')

  const suffix = output.endsWith('.nullable()') ? '.unwrap().and' : '.and'

  return (
    <Const
      export
      name={resolvedName}
      JSDoc={{
        comments: [description ? `@description ${transformers.jsStringEscape(description)}` : undefined].filter(Boolean),
      }}
    >
      {[
        output,
        keysToOmit?.length ? `${suffix}(z.object({ ${keysToOmit.map((key) => `${key}: z.never()`).join(',')} }))` : undefined,
        withTypeAnnotation && typeName ? ` as z.ZodType<${typeName}>` : '',
      ]
        .filter(Boolean)
        .join('') || ''}
    </Const>
  )
}

type FileProps = {}

Schema.File = function ({}: FileProps): ReactNode {
  const {
    pluginManager,
    plugin: {
      options: { typed },
    },
  } = useApp<PluginZod>()
  const { tree, schema } = useSchema()

  const withData = tree.some(
    (schema) =>
      schema.keyword === schemaKeywords.array ||
      schema.keyword === schemaKeywords.and ||
      schema.keyword === schemaKeywords.object ||
      schema.keyword === schemaKeywords.union ||
      schema.keyword === schemaKeywords.tuple,
  )

  const withTypeAnnotation = !!typed

  return (
    <Oas.Schema.File output={pluginManager.config.output.path}>
      <Schema.Imports />
      <File.Source>
        <Schema withTypeAnnotation={withTypeAnnotation} description={schema?.description} />
      </File.Source>
    </Oas.Schema.File>
  )
}
Schema.Imports = (): ReactNode => {
  const {
    pluginManager,
    plugin: {
      options: { typed, importPath },
    },
  } = useApp<PluginZod>()
  const { path: root } = useFile()
  const { name, tree, schema } = useSchema()

  // used for this.options.typed
  const typeName = pluginManager.resolveName({
    name,
    pluginKey: [pluginTsName],
    type: 'type',
  })

  const typeFileName = pluginManager.resolveName({
    name: name,
    pluginKey: [pluginTsName],
    type: 'file',
  })

  const typePath = pluginManager.resolvePath({
    baseName: typeFileName,
    pluginKey: [pluginTsName],
  })

  const withTypeAnnotation = !!typed

  return (
    <>
      <File.Import name={['z']} path={importPath} />
      {withTypeAnnotation && typeName && typePath && <File.Import isTypeOnly root={root} path={typePath} name={[typeName]} />}
    </>
  )
}
