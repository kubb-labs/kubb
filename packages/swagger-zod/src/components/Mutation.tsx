import { File, usePlugin, usePluginManager } from '@kubb/react'
import { useOas, useOperationFile, useSchemas } from '@kubb/swagger/hooks'

import { ZodBuilder } from '../ZodBuilder.ts'

import type { ReactNode } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type Props = {
  builder: ZodBuilder
}

export function Mutation({
  builder,
}: Props): ReactNode {
  const { source } = builder.build()

  return (
    <>
      {source}
    </>
  )
}

// eslint-disable-next-line @typescript-eslint/ban-types
type FileProps = {}

Mutation.File = function({}: FileProps): ReactNode {
  const { options } = usePlugin<PluginOptions>()

  const schemas = useSchemas()
  const pluginManager = usePluginManager()
  const oas = useOas()
  const file = useOperationFile()

  const builder = new ZodBuilder(options, { oas, pluginManager })
    .add(schemas.pathParams)
    .add(schemas.queryParams)
    .add(schemas.headerParams)
    .add(schemas.response)
    .add(schemas.request)
    .add(schemas.errors)

  const { source, imports } = builder.build()

  return (
    <>
      <File<FileMeta>
        baseName={file.baseName}
        path={file.path}
        meta={file.meta}
      >
        <File.Import name={['z']} path="zod" />
        {imports.map((item, index) => {
          return <File.Import key={index} root={file.path} {...item} />
        })}
        <File.Source>
          {source}
        </File.Source>
      </File>
    </>
  )
}
