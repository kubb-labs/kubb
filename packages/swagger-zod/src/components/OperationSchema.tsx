/* eslint-disable no-empty-pattern */
/* eslint-disable @typescript-eslint/ban-types */
import { Editor, File, usePlugin, usePluginManager } from '@kubb/react'
import { schemaKeywords } from '@kubb/swagger'
import { Oas, Schema } from '@kubb/swagger/components'
import { useGetOperationFile, useOas, useOperationSchemas } from '@kubb/swagger/hooks'

import { SchemaGenerator } from '../SchemaGenerator.tsx'

import type { KubbFile } from '@kubb/core'
import type { OperationSchema as OperationSchemaType } from '@kubb/swagger'
import type { ReactNode } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type Props = {}

export function OperationSchema({}: Props): ReactNode {
  return <></>
}

type FileProps = {
  mode: KubbFile.Mode | undefined
}

OperationSchema.File = function ({ mode = 'directory' }: FileProps): ReactNode {
  const plugin = usePlugin<PluginOptions>()

  const pluginManager = usePluginManager()
  const oas = useOas()
  const schemas = useOperationSchemas()
  const file = useGetOperationFile()

  const generator = new SchemaGenerator(plugin.options, {
    oas,
    plugin,
    pluginManager,
  })

  const items = [schemas.pathParams, schemas.queryParams, schemas.headerParams, schemas.statusCodes, schemas.request, schemas.response].flat().filter(Boolean)

  const mapItem = ({ name, schema: object, ...options }: OperationSchemaType, i: number) => {
    // hack so Params can be optional when needed
    const required = Array.isArray(object?.required) ? !!object.required.length : !!object?.required
    const optional = !required && !!name.includes('Params')

    return (
      <Oas.Schema key={i} generator={generator} name={name} object={object}>
        {mode === 'directory' && <Schema.Imports />}
        <File.Source>
          <Schema.Source extraSchemas={optional ? [{ keyword: schemaKeywords.optional }] : undefined} options={options} />
        </File.Source>
      </Oas.Schema>
    )
  }

  return (
    <Editor language="typescript">
      <File<FileMeta> baseName={file.baseName} path={file.path} meta={file.meta}>
        <File.Import name={['z']} path="zod" />
        {items.map(mapItem)}
      </File>
    </Editor>
  )
}
