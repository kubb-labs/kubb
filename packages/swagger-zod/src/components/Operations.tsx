import { Editor, File, usePlugin, usePluginManager } from '@kubb/react'
import { useOperationManager, useOperations } from '@kubb/swagger/hooks'

import transformers from '@kubb/core/transformers'
import type { KubbNode } from '@kubb/react'
import type { HttpMethod, Operation } from '@kubb/swagger/oas'
import type { ComponentProps, ComponentType } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  operationsName: string
  /**
   * Name of the function
   */
  pathsName: string
  operations: Operation[]
}

function Template({ operationsName, pathsName, operations }: TemplateProps): KubbNode {
  const { groupSchemasByByName } = useOperationManager()
  const transformedOperations = operations.map((operation) => ({ operation, data: groupSchemasByByName(operation, { type: 'function' }) }))

  const operationsJSON = transformedOperations.reduce(
    (prev, acc) => {
      prev[`"${acc.operation.getOperationId()}"`] = acc.data

      return prev
    },
    {} as Record<string, unknown>,
  )

  const pathsJSON = transformedOperations.reduce(
    (prev, acc) => {
      prev[`"${acc.operation.path}"`] = {
        ...(prev[`"${acc.operation.path}"`] || ({} as Record<HttpMethod, string>)),
        [acc.operation.method]: `operations["${acc.operation.getOperationId()}"]`,
      }

      return prev
    },
    {} as Record<string, Record<HttpMethod, string>>,
  )

  return (
    <>
      {`export const ${operationsName} = {
          ${transformers.stringifyObject(operationsJSON)}
       } as const;`}
      <br />
      {`export const ${pathsName} = {
         ${transformers.stringifyObject(pathsJSON)}
       } as const;`}
    </>
  )
}

type RootTemplateProps = {
  children?: React.ReactNode
}

function RootTemplate({ children }: RootTemplateProps) {
  const { key: pluginKey } = usePlugin<PluginOptions>()
  const pluginManager = usePluginManager()
  const { getFile } = useOperationManager()
  const operations = useOperations()
  const { groupSchemasByByName } = useOperationManager()
  const transformedOperations = operations.map((operation) => ({ operation, data: groupSchemasByByName(operation, { type: 'function' }) }))

  const file = pluginManager.getFile({ name: 'operations', extName: '.ts', pluginKey })
  const imports = Object.entries(transformedOperations)
    .map(([_key, { data, operation }], index) => {
      const names = [data.request, ...Object.values(data.responses), ...Object.values(data.parameters)].filter(Boolean)

      return <File.Import key={index} name={names} root={file.path} path={getFile(operation).path} />
    })
    .filter(Boolean)

  return (
    <Editor language="typescript">
      <File<FileMeta> baseName={file.baseName} path={file.path} meta={file.meta} exportable={false}>
        {imports}
        <File.Source>{children}</File.Source>
      </File>
    </Editor>
  )
}

const defaultTemplates = { default: Template, root: RootTemplate } as const

type Templates = Partial<typeof defaultTemplates>

type Props = {
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: ComponentType<ComponentProps<typeof Template>>
}

export function Operations({ Template = defaultTemplates.default }: Props): KubbNode {
  const operations = useOperations()

  return <Template operationsName="operations" pathsName="paths" operations={operations} />
}

type FileProps = {
  /**
   * This will make it possible to override the default behaviour.
   */
  templates?: Templates
}

Operations.File = function (props: FileProps): KubbNode {
  const templates = { ...defaultTemplates, ...props.templates }

  const Template = templates.default
  const RootTemplate = templates.root

  return (
    <RootTemplate>
      <Operations Template={Template} />
    </RootTemplate>
  )
}

Operations.templates = defaultTemplates
