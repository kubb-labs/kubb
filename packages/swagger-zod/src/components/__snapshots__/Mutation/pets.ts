export const CreatePets201 = z.any()
export const CreatePetsMutationRequest = z.object({ name: z.string().optional(), tag: z.string().optional() })
export const CreatePetsMutationResponse = z.any()
export const CreatePetsError = z.lazy(() => Error)
