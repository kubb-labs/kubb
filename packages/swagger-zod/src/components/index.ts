export { OperationSchema } from './OperationSchema.tsx'
