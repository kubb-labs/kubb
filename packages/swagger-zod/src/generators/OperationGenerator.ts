import { getRelativePath } from '@kubb/core/utils'
import { OperationGenerator as Generator, resolve } from '@kubb/swagger'

import { ZodBuilder } from '../builders/index.ts'

import type { KubbFile } from '@kubb/core'
import type { FileResolver, Operation, OperationSchemas, Resolver } from '@kubb/swagger'
import type { FileMeta, PluginOptions } from '../types.ts'

// eslint-disable-next-line @typescript-eslint/ban-types
type Options = {}

export class OperationGenerator extends Generator<Options, PluginOptions> {
  resolve(operation: Operation): Resolver {
    const { pluginManager, plugin } = this.context

    return resolve({
      operation,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
      pluginKey: plugin.key,
    })
  }

  async all(): Promise<KubbFile.File | null> {
    return null
  }

  async get(operation: Operation, schemas: OperationSchemas, _options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { pluginManager, mode, plugin, oas } = this.context

    const zod = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: zod.name, pluginKey: plugin.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new ZodBuilder({ fileResolver: mode === 'file' ? undefined : fileResolver, withJSDocs: true, resolveName: pluginManager.resolveName, oas })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    return {
      path: zod.path,
      baseName: zod.baseName,
      source,
      imports: [
        {
          name: ['z'],
          path: 'zod',
        },
      ],
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async post(operation: Operation, schemas: OperationSchemas, _options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { pluginManager, mode, plugin, oas } = this.context

    const zod = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: zod.name, pluginKey: plugin.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new ZodBuilder({ fileResolver: mode === 'file' ? undefined : fileResolver, withJSDocs: true, resolveName: pluginManager.resolveName, oas })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.request)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    return {
      path: zod.path,
      baseName: zod.baseName,
      source,
      imports: [
        {
          name: ['z'],
          path: 'zod',
        },
      ],
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async put(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async patch(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async delete(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
}
