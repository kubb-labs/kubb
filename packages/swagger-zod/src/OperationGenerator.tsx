import { createRoot } from '@kubb/react'
import { OperationGenerator as Generator } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OperationSchema } from './components/OperationSchema.tsx'

import type { KubbFile } from '@kubb/core'
import type { AppContextProps } from '@kubb/react'
import type { OperationMethodResult, OperationsByMethod } from '@kubb/swagger'
import type { Operation } from '@kubb/swagger/oas'
import { Operations } from './components/Operations.tsx'
import type { FileMeta, PluginOptions } from './types.ts'

export class OperationGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async all(operations: Operation[], _operationsByMethod: OperationsByMethod): OperationMethodResult<FileMeta> {
    const { pluginManager, oas, plugin } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    const templates = {
      operations: Operations.templates,
      ...this.options.templates,
    }

    root.render(
      <Oas oas={oas} operations={operations} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        {templates.operations && <Operations.File templates={templates.operations} />}
      </Oas>,
      { meta: { pluginManager, plugin } },
    )

    return root.files
  }

  async operation(operation: Operation, options: PluginOptions['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })
    root.render(
      <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        <Oas.Operation operation={operation}>
          <OperationSchema.File mode={mode} />
        </Oas.Operation>
      </Oas>,
      { meta: { pluginManager, plugin: { ...plugin, options } } },
    )

    return root.files
  }

  async get(): OperationMethodResult<FileMeta> {
    return null
  }

  async post(): OperationMethodResult<FileMeta> {
    return null
  }

  async put(): OperationMethodResult<FileMeta> {
    return null
  }
  async patch(): OperationMethodResult<FileMeta> {
    return null
  }
  async delete(): OperationMethodResult<FileMeta> {
    return null
  }
}
