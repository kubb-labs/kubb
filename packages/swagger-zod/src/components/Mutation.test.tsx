import { mockedPluginManager } from '@kubb/core/mocks'
import { createRootServer } from '@kubb/react/server'
import { OasManager } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OperationGenerator } from '../OperationGenerator.tsx'
import { Mutation } from './Mutation.tsx'

import type { Plugin } from '@kubb/core'
import type { AppContextProps } from '@kubb/react'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'
import type { PluginOptions } from '../types.ts'

describe('<Mutation/>', async () => {
  const oas = await OasManager.parseFromConfig({
    root: './',
    output: { path: 'test', clean: true },
    input: { path: 'packages/swagger-zod/mocks/petStore.yaml' },
  })

  const options: GetOperationGeneratorOptions<OperationGenerator> = {
    exclude: [],
    include: undefined,
    override: undefined,
    transformers: {},
    typed: false,
    dateType: 'string',
    unknownType: 'any',
  }

  const plugin = { options } as Plugin<PluginOptions>
  const og = await new OperationGenerator(
    options,
    {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin,
      contentType: undefined,
      override: undefined,
    },
  )

  test('pets', async () => {
    const operation = oas.operation('/pets', 'post')
    const context: AppContextProps<PluginOptions['appMeta']> = { meta: { pluginManager: mockedPluginManager, plugin } }

    const Component = () => {
      return (
        <Oas oas={oas} operations={[operation]} getSchemas={(...props) => og.getSchemas(...props)}>
          <Oas.Operation operation={operation}>
            <Mutation.File mode="directory" />
          </Oas.Operation>
        </Oas>
      )
    }

    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />, context)

    expect(output).toMatchFileSnapshot('./__snapshots__/Mutation/pets.ts')
  })
})
