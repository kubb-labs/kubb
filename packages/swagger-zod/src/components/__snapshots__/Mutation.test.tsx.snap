// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Mutation/> > pets 1`] = `
"/**
 * @description Null response
 */
export const CreatePets201 = z.any()
export const CreatePetsMutationRequest = z.object({ name: z.string(), tag: z.string() })
export const CreatePetsMutationResponse = z.any()

/**
 * @description unexpected error
 */
export const CreatePetsError = z.lazy(() => Error)
"
`;
